In [1]:
import sklearn

print(sklearn.__version__)

1.1.2


# 1. 필요한 모듈 import하기

In [2]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 2. 데이터 준비

### load_digits 메서드 사용

In [3]:
wine = load_wine()

# 3. 데이터 이해하기

### Feature Data 지정하기

In [4]:
wine_data = wine.data

### Label Data 지정하기

In [5]:
wine_label = wine.target
print(wine_label.shape)
wine_label

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

### Target Names 출력해 보기

In [6]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

### 데이터 Describe 해 보기

In [7]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            10.6  30.0    19.5   3.3
    Ma

# 4. train, test 데이터 분리

In [8]:
import pandas as pd

print(pd.__version__)

1.4.3


### 문제지

In [9]:
# 문제지
wine_df = pd.DataFrame(data = wine_data, columns=wine.feature_names)
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

### 정답지

In [10]:
# 정답지
wine_df["label"] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

### train, test 데이터 분리 

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size = 0.3,
                                                   random_state = 10)



print('X_train 개수: ', len(X_train),', X_test 개수: ', len(X_test))

X_train 개수:  124 , X_test 개수:  54


In [12]:
X_train.shape, y_train.shape

((124, 13), (124,))

In [13]:
X_test.shape, y_test.shape

((54, 13), (54,))

In [14]:
y_train, y_test

(array([2, 1, 0, 1, 2, 0, 0, 2, 2, 2, 2, 1, 1, 1, 2, 1, 0, 0, 1, 0, 0, 0,
        1, 2, 0, 0, 2, 0, 0, 0, 2, 2, 0, 2, 2, 0, 1, 2, 0, 1, 2, 2, 1, 2,
        0, 1, 0, 1, 0, 1, 1, 2, 2, 1, 0, 2, 1, 0, 2, 2, 1, 0, 0, 0, 1, 1,
        1, 2, 0, 1, 0, 0, 0, 0, 2, 1, 1, 2, 2, 1, 1, 2, 1, 0, 0, 0, 0, 0,
        1, 2, 1, 0, 1, 1, 2, 1, 0, 1, 2, 2, 0, 1, 1, 1, 1, 0, 2, 1, 0, 1,
        2, 0, 0, 1, 0, 2, 2, 2, 1, 1, 1, 0, 1, 0]),
 array([1, 1, 0, 1, 0, 1, 1, 0, 2, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 2, 0, 2,
        0, 0, 1, 2, 1, 2, 1, 1, 2, 2, 1, 2, 1, 0, 1, 1, 0, 1, 1, 0, 2, 1,
        0, 1, 2, 1, 1, 0, 2, 1, 2, 0]))

# 5. 다양한 모델로 학습시켜보기

### Decision Tree

In [15]:
# Decision Tree
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
print(decision_tree._estimator_type)


classifier


In [16]:
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

In [17]:
y_pred = decision_tree.predict(X_test) # test 데이터로 예측
                                        # X_test데이터에는 정답인 label이 없고 
                                        # feature데이터만 존재
                                        # decision_tree 에 X_test 데이터로 predict
                                        # 실행하면 모델이 예측한 y_pred 도출
y_pred

array([1, 1, 0, 1, 0, 1, 1, 0, 2, 0, 0, 0, 0, 2, 1, 1, 2, 1, 1, 2, 0, 2,
       0, 0, 2, 1, 1, 2, 1, 1, 2, 2, 2, 2, 1, 0, 0, 1, 0, 1, 0, 0, 2, 0,
       0, 1, 2, 1, 1, 0, 2, 1, 2, 0])

In [18]:
y_test

array([1, 1, 0, 1, 0, 1, 1, 0, 2, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 2, 0, 2,
       0, 0, 1, 2, 1, 2, 1, 1, 2, 2, 1, 2, 1, 0, 1, 1, 0, 1, 1, 0, 2, 1,
       0, 1, 2, 1, 1, 0, 2, 1, 2, 0])

In [19]:
# 정확도
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.8333333333333334

### Random Forest 

In [20]:
# Random Forest 

from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size = 0.3,
                                                   random_state = 10)

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      0.93      0.96        27
           2       0.86      1.00      0.92        12

    accuracy                           0.96        54
   macro avg       0.95      0.98      0.96        54
weighted avg       0.97      0.96      0.96        54



In [21]:
# 정확도
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.9629629629629629

### SVM

In [22]:
# SVM
from sklearn import svm
svm_model = svm.SVC()

print(svm_model._estimator_type)

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       0.75      0.60      0.67        15
           1       0.81      0.63      0.71        27
           2       0.33      0.58      0.42        12

    accuracy                           0.61        54
   macro avg       0.63      0.60      0.60        54
weighted avg       0.69      0.61      0.63        54



In [23]:
# 정확도
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.6111111111111112

### SGD Classifier

In [24]:
# SGD Classifier
from sklearn.linear_model import SGDClassifier
import warnings 

#warnings.filterwarnings('always') # 항상 경고 표시
warnings.filterwarnings('ignore') # 발생하는 경고무시

sgd_model = SGDClassifier()

print(sgd_model._estimator_type)


sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       1.00      0.33      0.50        15
           1       0.75      0.11      0.19        27
           2       0.24      0.92      0.39        12

    accuracy                           0.35        54
   macro avg       0.66      0.45      0.36        54
weighted avg       0.71      0.35      0.32        54



In [25]:
# 정확도
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.35185185185185186

### Logistic Regression

In [26]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=4000)

print(logistic_model._estimator_type)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)

print(classification_report(y_test, y_pred))

classifier
              precision    recall  f1-score   support

           0       0.78      0.93      0.85        15
           1       0.95      0.78      0.86        27
           2       0.86      1.00      0.92        12

    accuracy                           0.87        54
   macro avg       0.86      0.90      0.88        54
weighted avg       0.88      0.87      0.87        54



In [27]:
# 정확도
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
accuracy

0.8703703703703703

# 6. 모델을 평가해보기

#### Decision Tree 의 정확도 :        0.8333333333333334

#### Random Forest 의 정확도 :        0.9629629629629629

#### SVM 의 정확도 :                  0.6111111111111112

#### SGD Classifier 의 정확도 :       0.6851851851851852

#### Logistic Regression  의 정확도 : 0.8148148148148148

## 예측결과==============

많은 요소를 고려해 트리를 만들어 예측하는 Random Forest 모델이 적합하다고 생각

의사결정트리도 좋지만 위에서 정확도를 보고

와인 데이터의 요소가 많으니 random forest 모델로 예측하는것이 제일 좋지 않을까.. 하는 생각

## =====================

### 오차 행렬

In [28]:
#random forest

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

X_train, X_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size = 0.3,
                                                   random_state = 10)

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

# 정확도
accuracy = accuracy_score(y_test, y_pred)
accuracy

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      0.93      0.96        27
           2       0.86      1.00      0.92        12

    accuracy                           0.96        54
   macro avg       0.95      0.98      0.96        54
weighted avg       0.97      0.96      0.96        54



0.9629629629629629

In [29]:
# 오차 행렬
from sklearn.metrics import confusion_matrix


confusion_matrix(y_test, y_pred)

array([[15,  0,  0],
       [ 0, 25,  2],
       [ 0,  0, 12]], dtype=int64)

In [30]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      0.93      0.96        27
           2       0.86      1.00      0.92        12

    accuracy                           0.96        54
   macro avg       0.95      0.98      0.96        54
weighted avg       0.97      0.96      0.96        54



## 성능을 평가하는 지표로는 무엇이 좋은지?
## 평가지표 중 적절한것
## 선택의 이유는?

Accuracy로 평가하는게 제일 적절하다고 생각

전체 대비 정확히 예측한 개수의 비율

즉, 전체 개수 중 맞은 것의 개수를 나타내기때문

예측값과 실제값이 얼마나 동일한가에 대한 비율 만으로 결정됨 

-> 오차행렬에서 True에 해당하는 TN(거짓 양성)과 TP(참 양성)에 의해서만 좌우된다


***
precision도 recall의 조화평균인 f1-score가 적절하다고 판한한다

얼마나 잘 예측했는지 알려주는 precision도 좋을 것 같다